### ISSUE TITLE
BUG: crash / core dump when calling scipy.stats.beta.ppf with certain parameters (boost-related)

### ISSUE THREAD

**Issue by pshargo** [scipy/scipy#14606]
### Reproducing Code Example

```python
import scipy.stats
scipy.stats.beta.ppf(0.995, 1.0e11, 1.0e13)
```

### Error message

```shell
terminate called after throwing an instance of 'boost::wrapexcept<boost::math::evaluation_error>'
  what():  Error in function boost::math::tgamma<double>(double,double): Series evaluation exceeded 1000000 iterations, giving up now.
Aborted (core dumped)
```

### SciPy/NumPy/Python version information

1.7.1 1.21.2 sys.version_info(major=3, minor=9, micro=6, releaselevel='final', serial=0)

**Comment by pshargo** [comment]
Note that this crash only occurs for scipy v1.7.0+, which is when the boost versions of many stats functions (including for the beta distribution) were first integrated. My suspicion is that scipy's boost error-handling policies just aren't catching this particular exception.

**Comment by mdhaber** [comment]
Sounds like an issue in underlying Boost code. @mckib2 are you able to check this easily?

But yes, sounds like we should return a Python error in this case.

**Comment by pshargo** [comment]
With the (large!) caveat that I don't know much about this stuff, the fact that the error message references `boost::math::evaluation_error` suggests to me that perhaps `('BOOST_MATH_EVALUATION_ERROR_POLICY', 'ignore_error')` should be added to the list of `DEFINES` in setup.py. In fact, there are a bunch of `BOOST_MATH_xxx_POLICY` values [defined in the boost repo](https://github.com/boostorg/math/blob/b2538faaf9802af8e856c603b9001e33db826676/include/boost/math/tools/user.hpp); would it make sense to include most/all of them as a defensive measure?

In any event, I really appreciate you all taking a look at this!

**Comment by mckib2** [comment]
> Sounds like an issue in underlying Boost code. @mckib2 are you able to check this easily?

Checking it out right now, I agree we should try our best to make this a Python exception at the very least

**Comment by mckib2** [comment]
Draft PR: gh-14618

It turns out you can return the best guess in this scenario, so I wrote a function to handle the `evaluation_error` by raising a warning and returning whatever value is available.  @mdhaber Is there a better warning type to use instead of `RuntimeWarning`?

**Comment by mdhaber** [comment]
I'm not aware of a more specific one.

**Comment by pshargo** [comment]
IMO the draft PR's approach of raising a `RuntimeWarning` and returning whatever value is available (whether or not it's legitimate) would be the preferred approach. That way the user could either run with the returned value as-is, or catch it with the `warnings` library and do whatever special handling is preferable for their use case.

I don't really have the expertise to provide detailed feedback on @mckib2 's draft PR, but to my eyes it looks like it would handle this particular issue in an appropriate way. Have either of you put any thought into whether or not it would make sense to add more of the boost math policies [listed here](https://github.com/boostorg/math/blob/b2538faaf9802af8e856c603b9001e33db826676/include/boost/math/tools/user.hpp) to the list of `DEFINES` in `scipy/stats/_boost/setup.py`? There appear to be 2 others (`BOOST_MATH_POLE_ERROR_POLICY ` and `BOOST_MATH_OVERFLOW_ERROR_POLICY `) that default to `throw_on_error`, which I'm guessing would cause crashes just like `BOOST_MATH_EVALUATION_ERROR_POLICY` did.

Thanks again for looking into this!

**Comment by mckib2** [comment]
I am not sure if any of the functions we use will ever raise a pole error (some special functions like digamma might), so I'm not sure how to create a test that would raise that error.  There's already a `RuntimeWarning` being raised for overflow errors encountered like this:
```python
scipy.stats.beta.ppf(.1, 1e10, 1e20)
```
I've played around a little bit, but again I'm not sure how to go about getting a Boost `overflow_error`.  If you're able to provide a producer for those errors it is straightforward to raise a `RuntimeWarning` or error like I've done in the PR, otherwise it might be best to wait until someone encounters these errors so we know what tests to write

**Comment by pshargo** [comment]
Yep, good point re: the need for test cases. I don't have any producers for those error types at the ready, so I'm in agreement that at this point it's probably best to wait until someone encounters them. It's very possible that won't happen until/unless `scipy.stats` starts pulling in more distributions from boost.
